Dage 21: Vespa Search Engine

Intrinsical-AI opened the pull request:
# [Task: DAGE-21](https://sease.atlassian.net/browse/DAGE-21?atlOrigin=eyJpIjoiODA5OTE4YmE3NzE5NDY2ODg3ZjMyNmMwNTRiNGVhMzkiLCJwIjoiaiJ9)

## Context
[Previous PR](https://github.com/SeaseLtd/rated-ranking-evaluator/pull/190)

## (features): vespa-search-engine 
- Fix proposal over the overriding fl params in the query template (found bug in Solr)
- Python class to perform search over a running instance
- Containerized (Compose + Docker + Init script) Vespa instance with document feeding.
- Bash script, intended to be run from the container; vespa init orchestration + healtchecks / validation + doc feeding
- Aim for a robust approach using quotes to avoid injection
- Added updated tests

nicolo-rinaldi commented on rre-tools/dataset-generator/tests/unit/test_vespa_search_engine.py:
```diff
@@ -0,0 +1,251 @@
+from dataset_generator.config import Config
+from dataset_generator.search_engine import VespaSearchEngine
+from commons.model.document import Document  # <- usa el mismo import que en la implementación
+from dataset_generator.utils import clean_text  # para expectations del normalizador
```
Remember to change Spanish comments with English ones

nicolo-rinaldi reviewed (commented):
You didn't update the factory and in the config there is a new param called `vespa_schema`. How are we dealing with those in the main script?

I will end the review tomorrow

nicolo-rinaldi commented on rre-tools/docker-services/vespa-init/app/vespa-init.sh:
Do we need this empty file?

nicolo-rinaldi commented on rre-tools/docker-services/docker-compose.vespa.yml:
```diff
@@ -0,0 +1,27 @@
+    volumes:
+      - ./vespa-init/app:/app
+      - ./vespa-init/data:/data
+      - ./vespa-init/vespa-init.sh:/app/vespa-init.sh
```
I saw that the file is also here. Is it needed to let vespa work?

Intrinsical-AI replied:
Yes

nicolo-rinaldi commented on rre-tools/docker-services/vespa-init/vespa-init.sh:
```diff
@@ -0,0 +1,69 @@
+done
+
+# Feed (opt)
+EXPECTED_DOCS=$(ls /data/*.json 2>/dev/null | wc -l | tr -d ' ')
```
I'd rather call this `NUM_EXPECTED_DOCS`, just because I took a while to understand that it was a number

Intrinsical-AI replied:
Good point. I will change it and encapsulate into functions to make it easier to read

nicolo-rinaldi commented on rre-tools/dataset-generator/tests/unit/mocks/vespa.py:
Why do we need these mocks if we are not using them in the tests? I might be missing something

Intrinsical-AI replied:
You're right, I was using it in the old PR tests. I will re-introduce it in current ones

Intrinsical-AI committed:
feat(vespa): Add Vespa search engine integration

- Add VespaSearchEngine with userInput() parameter substitution
- Add comprehensive test suite with mocks and resources
- Add Docker Compose setup with vespa-init service
- Add YQL templates for search queries
- Update Config to support vespa_schema parameter
- Update SearchEngineFactory to register Vespa engine
- Add documentation for Vespa Docker services

Intrinsical-AI commented:
I've a problem due to conflicts when pushing the last version. 

- Rebased feature branch onto latest dataset-generator
- Resolved add/add conflicts by keeping feature changes
- Force pushed with --force-with-lease to update PR clean

Intrinsical-AI commented on rre-tools/dataset-generator/src/dataset_generator/config.py:
```diff
@@ -120,6 +128,12 @@ def check_rre_fields_required(self) -> "Config":
+    @model_validator(mode="after")
+    def check_vespa_fields_required(self) -> "Config":
+        if self.search_engine_type == "vespa" and not self.vespa_schema:
+            raise ValueError("vespa_schema is required when search_engine_type='vespa'")
```
Could be changed to log the missing param in conf + fallback to self.vespa_schema = "doc" as default

nicolo-rinaldi replied:
I think it's fine raising an error here

Intrinsical-AI commented on rre-tools/dataset-generator/tests/unit/resources/vespa_good_config.yaml:
```diff
@@ -0,0 +1,22 @@
+search_engine_type: "vespa"
+collection_name: "doc"
+vespa_schema: "doc"
```
New (optional) parameter for the config

Intrinsical-AI commented on rre-tools/dataset-generator/src/dataset_generator/config.py:
```diff
@@ -17,6 +17,7 @@ class Config(BaseModel):
     )
     search_engine_type: Literal['solr', 'elasticsearch', 'opensearch', 'vespa']
     collection_name: str = Field(..., description="Name of the index/collection of the search engine")
+    vespa_schema: Optional[str] = Field(None, description="Schema name for Vespa search engine")
```
Parameter specfic for vespa schema (check docker_services/vespa-init/app) -> Optional (we can simply avoid it from conf). No breaking changes, still working backwards.

Intrinsical-AI commented on rre-tools/docker-services/vespa-init/data/dataset.json:
```diff
@@ -0,0 +1,14 @@
+[
+  {
+    "put": "id:doc:doc::1",
```
Vespa expects this format / syntax for data feeding

nicolo-rinaldi commented on rre-tools/dataset-generator/tests/unit/test_vespa_search_engine.py:
```diff
@@ -0,0 +1,251 @@
+from dataset_generator.config import Config
+from dataset_generator.search_engine import VespaSearchEngine
+from commons.model.document import Document  # <- usa el mismo import que en la implementación
+from commons.utils import clean_text  # para expectations del normalizador
```
There are still Spanish comments here

Intrinsical-AI replied:
Pretty sure these are the same I forgot to remove haha (cleaning incoming)

nicolo-rinaldi commented on VESPA-CHEETSHEET.md:
```diff
@@ -0,0 +1,99 @@
+
+## 1. Environment Management (Makefile)
+
+These commands should be run from the `tests/integration/` directory. The `Makefile` is the simplest way to manage the container lifecycle.
```
I cannot see this Makefile

Intrinsical-AI replied:
Old doc / legacy. The commands still useful. I should remove it

nicolo-rinaldi commented on VESPA-CHEETSHEET.md:
```diff
@@ -0,0 +1,99 @@
+
+---
+
+## 5. Running Integration Tests
```
And I don't find this either

nicolo-rinaldi commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/vespa_search_engine.py:
```diff
@@ -0,0 +1,297 @@
+class VespaSearchEngine(BaseSearchEngine):
+    """
+    Thin HTTP wrapper around the Vespa Query API.
+    Assumes an already deployed a schema called `doc`.
```
Now we are dealing with `doc` as a parameter in config

Intrinsical-AI replied:
What do you mean? Not getting you exactly. I added a param in the config called vespa_schema (?)

nicolo-rinaldi commented on rre-tools/dataset-generator/src/dataset_generator/search_engine/vespa_search_engine.py:
```diff
@@ -0,0 +1,297 @@
+            "hits": int(doc_number),
+            "presentation.format": "json",
+        }
+        log.debug(f"Vespa payload (generation): {str(payload)[:1000]}")
```
Why are we capping at 1k characters?

Intrinsical-AI replied:
I think the payload might contain the full document, or very large vectors (let's say > 1000 dimensional). Just to not flood the logs. Maybe I should explicitly log that also. E.g:

`log.debug(f"Vespa payload (showing 1000 first chars): {str(payload)[:1000]}")`

Intrinsical-AI replied:
Done (added explicitly to the debug - log)

Intrinsical-AI committed:
Merge branch 'dataset-generator' of https://github.com/SeaseLtd/rated-ranking-evaluator into DAGE-21/vespa-search-engine-clean

Intrinsical-AI committed:
(clean): following Nicolos feedback to clean comments and logs. General coherence and consistancy

Intrinsical-AI committed:
(clean): following Nicolos feedback to clean comments and logs. General coherence and consistancy

Intrinsical-AI committed:
(clean): following Nicolos feedback to clean comments and logs. General coherence and consistancy
